cmd/kube-controller-manager/app/plugins.go

```diff
@@ -66,9 +70,50 @@ func ProbeRecyclableVolumePlugins(flags VolumeConfigFlags) []volume.VolumePlugin
 	}
 	allPlugins = append(allPlugins, nfs.ProbeVolumePlugins(nfsConfig)...)
 
+	allPlugins = append(allPlugins, aws_ebs.ProbeVolumePlugins()...)
+	allPlugins = append(allPlugins, gce_pd.ProbeVolumePlugins()...)
+	allPlugins = append(allPlugins, cinder.ProbeVolumePlugins()...)
+
 	return allPlugins
 }
 
+// NewVolumeProvisionersForCloud maps a cloud provider to a specific volume plugin.
+func NewVolumeProvisioners(plugins []volume.VolumePlugin, qosClasses []string) map[string]volume.ProvisionableVolumePlugin {
+	provisioners := map[string]volume.ProvisionableVolumePlugin{}
+	for _, qos := range qosClasses {
+		// the value is "key/value" and requires parsing after the first slash.
+		// values will be volume plugin names, many of which also contain slashes in the name.
+		firstSlash := strings.Index(qos, "/")
+		// 0 cannot be the first slash. there would be no tier name.
+		if firstSlash > 0 {
+			qosClass := qos[0:firstSlash]
+			// add the QoS class to the provisioner map but missing the provisioner.
+			// a suitable plugin should be found and added to the map.
+			// we get to raise an error if the provisioner remains nil
+			provisioners[qosClass] = nil
+			provisioner := qos[(firstSlash + 1):] // needs +1 because we don't want the slash prefix
+			for _, plugin := range plugins {
+				if plugin.Name() == provisioner {
+					if provisonablePlugin, ok := plugin.(volume.ProvisionableVolumePlugin); ok {
+						provisioners[qosClass] = provisonablePlugin
+					}
+				}
+			}
+		}
+	}
+	anyNil := false
+	for qosClass, plugin := range provisioners {
+		if plugin == nil {
+			anyNil = true
+			glog.Warningf("Could not find a ProvisionableVolumePlugin for QoS class %s", qosClass)
+		}
+	}
+	if anyNil {
+		glog.Fatalf("One or more QoS Classes were incorrectly configured and are missing a provisioner")
```

Comment:
return an error (or error list if you want to aggregate so you can give info on exactly what is configured wrong) and let the caller fatal it if they want.